### DISCUSSION TITLE
Pre-caching remote video streams from a CMS

### DISCUSSION THREAD

**Discussion by scottfuoco** [serwist/serwist#53] [Q&A]
Hey there, first thanks for creating a maintaining this project!  I'm fairly new to PWAs and this is really the only resource I can find for creating them within a Nextjs project so truly thanks again.  If I can, I'll start with a bit of a description of my project, and then go into the PWA features we are attempting to add.

I have a single page interactive application (no url or route changes) written with Nextjs.  Users can click on objects they will expand to show a carousel with images, videos and a text description within them.  Currently the assets within these expanded objects are fetched from our CMS, which uses AWS S3, when the objects are clicked on, so not when the application initially loads.

I'm having difficulty creating an installable PWA that has the following features and I'm wondering if you could guide me in the right direction for accomplishing these desired features or let me know if it isn't possible 🥲

Our goal is to have offline support and where hoping to have the following when a user installs the PWA:
- download and cache all the content and assets for the application, including those that get loaded after interacting with the objects on the page
- if the network is available use that for requests and fall back to the cache if offline
- if a fetch from the network is made update the cached version of that request
- periodically try to update the cache in the background (nice to have)

Some challenges and unknowns we have are:
- how to pre-cache all assets on the installation of the PWA, including the ones that aren't loaded until the user clicks on the objects and scrolls through the carousel
- how to download and cache the video files that are streamed, and not downloaded in full like images
- we are confused about where this configuration would live.  Is this where we would create a custom service worker, install google workbox, and add custom cache strategies? or is there another approach we should be taking?

Thanks again in advance!

**Comment by DuCanhGH** [comment]
@scottfuoco 

Note: some features mentioned here have not been documented yet. For documentation on such features, see Workbox's docs or read the JSDoc if that is not available. Sorry for the inconvenience.

> if the network is available use that for requests and fall back to the cache if offline
> if a fetch from the network is made update the cached version of that request

You can use `runtimeCaching` through `@serwist/sw.registerRuntimeCaching`, [`@serwist/sw.installSerwist`](https://serwist.pages.dev/docs/sw/install-serwist), or [`@serwist/routing.registerRoute`](https://developer.chrome.com/docs/workbox/modules/workbox-routing#method-registerRoute) with [`@serwist/strategies.NetworkFirst`](https://developer.chrome.com/docs/workbox/modules/workbox-strategies#type-NetworkFirst). 

> periodically try to update the cache in the background (nice to have)

Hmm, you can make use of [the Web Periodic Background Synchronization API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Periodic_Background_Synchronization_API). Do note that it isn't supported on Firefox and Safari and that the implementation... can get quite verbose.

> how to pre-cache all assets on the installation of the PWA, including the ones that aren't loaded until the user clicks on the objects and scrolls through the carousel

Depends on the assets you are trying to precache. Static Next.js JavaScript files are automatically precached. So are files in the public directory if you don't have [`additionalPrecacheEntries`](https://serwist.pages.dev/docs/build/configuring/base-options/additional-precache-entries) defined. Others aren't, but you can manually add them to `additionalPrecacheEntries` (again, remember to add files in the public directory if you do). Note that you have to find a way to predetermine a list of additional precache entries. Really unfortunate but there isn't an alternative...

> how to download and cache the video files that are streamed, and not downloaded in full like images

Use [`@serwist/range-requests`](https://developer.chrome.com/docs/workbox/modules/workbox-range-requests). If you use `runtimeCaching`, simply set `options.rangeRequests` to `true` for the entry that matches your streamed videos.

>  Is this where we would create a custom service worker, install google workbox, and add custom cache strategies? or is there another approach we should be taking?

This IS a Workbox fork and alternative. Just clarifying :)

To use Serwist, you always need a service worker source. If you are asking for something like Workbox's GenerateSW, there's not something like that, but `@serwist/sw.installSerwist` serves as an alternative. The strategies provided by `@serwist/strategies` should generally suits a handful of use cases, but if yours doesn't work nicely with them, you will have to write your own strategy.

Hope this helps!

**Reply by scottfuoco** [reply]
Awesome thanks for your help!  I look forward to playing around with these options / suggestions.
